text/0000-export-to-oci.md

```diff
@@ -0,0 +1,715 @@
+    - I didn't use the [skopeo]() in these cases to complete the pushing into the Daemon
+
+  Here are the results:
+
+    ![](https://i.imgur.com/vtOjxJP.png)
+
+     -  The results are actually very similar to exporting to OCI layout format for Java and Kotlin, but Ruby application is actually worst.
+
+    ![](https://i.imgur.com/FfbqfF6.png)
+
+     - Second build is actually better compared with the export to OCI in disk, Java and Kotlin increases the time just by **2%**, but Ruby again is worst
+
+  Some thoughts about this approaches
+
+    - **Process Management:** Platforms must now manage a parallel process (registry in the daemon). This would entail ensuring that the registry is started and cleaned up appropriately.
+    - **Networking:** There are additional network complications in order to route images to the ephemeral registry. For example, [network drivers](https://docs.docker.com/network/#network-drivers), [proxy](https://docs.docker.com/desktop/networking/#httphttps-proxy-support) and [DNS configuration](https://docs.docker.com/config/containers/container-networking/#dns-services), [host name resolution](https://docs.docker.com/desktop/networking/#i-want-to-connect-from-a-container-to-a-service-on-the-host), and [TLS certificates](https://betterprogramming.pub/deploy-a-docker-registry-using-tls-and-htpasswd-56dd57a1215a) to name a few.
+
+- Why is this proposal the best? [OCI Image Layout](https://github.com/opencontainers/image-spec/blob/main/image-layout.md) format is a standard from which other tools can create a [OCI Runtime Specification bundle](https://github.com/opencontainers/runtime-spec/blob/v1.0.0/bundle.md) exporting to this format enables Platforms to implement any feature in the top of this format, for example, exporting to [containerd](https://containerd.io) has been [requested](https://github.com/buildpacks/lifecycle/issues/829) by the community and it could be implemented if we can get the application image exported in [OCI Image Layout](https://github.com/opencontainers/image-spec/blob/main/image-layout.md) format.
+- What is the impact of not doing this? Probably will never remove the Daemon support in the Lifecycle
+
+# Prior Art
+[prior-art]: #prior-art
+
+- Discussion around removing the Daemon support [RFC](https://github.com/buildpacks/rfcs/blob/jjbustamante/feature/deprecate-daemon/text/0000-deprecate-daemon.md)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to be resolved before this gets merged?
+  - Tools like [umoci](https://umo.ci/) used to create a runtime bundle from an image in OCI layout format, requires the [annotation](https://github.com/opencontainers/image-spec/blob/main/annotations.md#pre-defined-annotation-keys) `org.opencontainers.image.ref.name` to be present. Also tools like [skopeo](https://github.com/containers/skopeo) when an image is `copy` in oci format the annotation is included.
```

Comment:
My vote would be to write this annotation. It feels weird to not be able to interact with other oci layout tooling.

Reply:
Would be fine to keep that as implementation detail and do not add it to the RFC ?

Reply:
This would introduce a slight difference between layout images and remote images. Is that an acceptable difference? I'm not aware of any others.

Reply:
I created a [PR](https://github.com/buildpacks/imgutil/pull/175)  to add it

Reply:
Would it be reasonable to add this to remote images as well?

Reply:
FWIW, `crane pull <image> <path> --format=oci` doesn't add the annotation... I don't know what to make of that.

Reply:
It does, if you add ` --annotate-ref` flag

Reply:
When exporting to layout we can derive `org.opencontainers.image.ref.name` from the target path, but maybe we could also add an optional `-annotate-ref` (or something) to the lifecycle so that when publishing to the registry we could optionally add the annotation if users want it. This would allow us to keep parity for remote & layout images which seems desirable.

Reply:
@natalieparellano I have a couple of things regarding the way the annotation is going to be implemented. 
1. `creator` already have a `tag` flag which works in this case.
2. I am not 100% on deriving the *tag* from the path. For example, if I am pack, I will use the `layout` package exposed on imgutil to download the `run-image` and saving it on disk before calling the lifecycle. Pack can save the `run-image` on whatever path it wants but if I try to determine the `org.opencontainers.image.ref.name` base on the path I could be tagging wrong the image.

Suppose the `run-image` is `ubuntu:bionic` and pack wants to save the image on `/home/user/.pack/repo/ubuntu`. Pack could create the image as follows:

``` =go
# remote-image is v1.Image load from remote `ubuntu:bionic`
image, _ := layout.NewImage("/home/user/.pack/repo/ubuntu", layout.FromBaseImage(remote-image))
image.Save()
```
If we try to determine the `org.opencontainers.image.ref.name` from the path, we could tagging the image with 
``` =go
org.opencontainers.image.ref.name: "ubuntu"
```
**Proposed solution**

I think I prefer to delete the tagging when you call the `Save(additionalNames ...string)` method, so on my previous example:

#### example 1

```=golang
# remote-image is v1.Image load from remote `ubuntu:bionic`
image, _ := layout.NewImage("/home/user/.pack/repo/ubuntu", layout.FromBaseImage(remote-image))
image.Save()
```

output: 

No `org.opencontainers.image.ref.name` is added because we didn't pass any additional name

#### example 2

```=golang
# remote-image is v1.Image load from remote `ubuntu:bionic`
image, _ := layout.NewImage("/home/user/.pack/repo/ubuntu", layout.FromBaseImage(remote-image))
image.Save("bionic")
```

output:

```=json
org.opencontainers.image.ref.name: "bionic"
```

Reply:
This is just an opinion, but it is confusing to me that we are overloading the `Save` method in this way. It feels like `Save` has side effects and requires library consumers to know too much.

For the remote package, image.Save("other name") means "save my image and also put it here". The creator treats `-tag` this way (i.e., `-tag some.registry.io/my-image/my-tag` and not `-tag my-tag`. In the layout package, I would expect that image.Save("other path") would also put the image at "other path".

Reply:
What I would propose instead is a new method on the Image interface like Image.AnnotateRef("my-tag") that the lifecycle and others could call instead as needed.